### PR TITLE
fix(ferment): close the proxy gap — charter audit, evidence hierarchy, judge observability

### PR DESCRIPTION
**Pull request by mikenorgate**
## Linked issue

Deferred per maintainer direction.

## What does this PR do?

Closes the **proxy gap** surfaced by an A/B comparison of two sessions that built the same Tahoe-desktop web app: a ferment session (`019fd7a2`) vs a singleshot session (`019fd6e5`). The ferment session scored higher on script-decidable gates but shipped a worse artifact — it optimized the proxy criteria the pipeline measured rather than the artifact judged in its native medium. This branch **widens what ferment evaluates** (never how work is produced), staying task-generic.

### v1 — widen the evaluation surface

| Chunk | Commit | What |
|-------|--------|------|
| `inspected` verify class | `250752bb` | New S2 verification vocabulary for one-off native-medium observations (run the artifact, record what it showed) — fills the gap between smoke/test (script-decidable) and walkthrough (narrated). |
| Intent charter | `fef646e9` | `FermentCharter` (intent + wow-factor + demo-script + constraints) persisted and re-injected at every anchor point (scoping, planning, phase/step grading, completion). Two-renderer budget keeps charter cost bounded. |
| Compaction hardening | `9c7344b4` | Open-concerns channel in compaction summaries + `isDegenerateCompactionSummary` breadcrumb — charter and quality signals survive context compression. |
| Advisory scoping signals | `f80a71fa` | `self_critique`, `scope_deltas`, `constraint_costs`, `quality_dimensions` on `ProposeScopingParams`/`ScopeParams` — surfaced to the agent as advisory sections (never persisted as gates). |
| Grader walkthrough evidence | (Chunk 4) | `artifact_evidence` (text/DOM first, image behind capability detection) + white-box grader guidance. |
| Ship-level charter audit | `996b261c` | `CharterClauseVerdict {clause, status: met\|waived\|unmet, evidence}` on `JudgeGrade` — per-clause verdicts at completion. |

### v2 — enforce + observe (from replay learnings)

| Chunk | Commit | What |
|-------|--------|------|
| A — charter-audit enforcement | `77eb3d71` | `charter_verdicts` REQUIRED in the journey-grader contract when a charter exists; single-shot retries up to the grade budget, then **soft-degrades** with a breadcrumb (never blocks ship). Subagent graders fall back to single-shot when the audit is missing. |
| B — evidence hierarchy | `04357371` | S2 gate guidance states the preference explicitly: executable checks (test > smoke) > walkthrough evidence (command + observed output) > one-off `inspected`. |
| C — judge observability | `9dc21af4` | `JudgeGrade.gradedBy` records the actual judge model (`describeJudgeModel()` mirrors `judgeApiCall` resolution); persisted on ferment + phase grades, rendered at completion (`· judge: provider/id`). The replay's same-model circularity would now be visible post-hoc. |
| D — harness hardening | `5b6c8c61` | Acceptance harvester writes `<out>/crash.log` on any failure; default probe reports `windowSurfaces`/`openDialogs` rest-state inventories. |

### Acceptance evidence

Baseline harvested for old-ferment vs singleshot vs post-fix replay. Post-fix replay graded **A** with execution-anchored evidence (126 unit + 99 playwright tests, `getComputedStyle` backdrop-filter, reload persistence, axe-core clean). The `charter_verdicts` enforcement (Chunk A) was not exercised by that replay (pre-v2 run) — it's the one open item for the next replay.

### Design constraints honored

- **Additive schemas only** — in-flight/resumed ferments never break; all new fields are optional.
- **No new blocking gates in v1** — every new signal ships as vocabulary, advisory flag, evidence field, or report structure.
- **Chunk A enforcement is soft** — retry bounded by `JOURNEY_GRADE_MAX_ATTEMPTS=3`, then grade-without-audit + breadcrumb; never fails ship on judge flakiness.
- **Task-generic** — no UI-only special-casing; fixes widen what is *evaluated*, never mandate *how* work is produced.

## Checklist

- [x] Tests pass locally (`pnpm run test`) — 569 files / 11,500 tests green
- [x] Lint passes (`pnpm run check`) — tsc clean, biome clean (135 files)
- [x] Documentation updated if behavior changed — guidance is in-product (gate-registry, prompt-block); no external docs changed
- [ ] This PR links to an open issue above — deferred per maintainer direction

### CODE REVIEW

File: scripts/benchmark/ferment-acceptance.mjs [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+	const page = await browser.newPage({ viewport: { width: 1440, height: 900 } })
+	await page.addInitScript(() => {
+		window.__harvestErrors = []
+		window.addEventListener("error", (e) => {
+			window.__harvestErrors.push(String(e.message ?? e))
+		})
+		console.error = new Proxy(console.error, {
+			apply(target, thisArg, argv) {
+				window.__harvestErrors.push(String(argv[0] ?? ""))
+				return Reflect.apply(target, thisArg, argv)
```
[the remark this answers is not in the document]
Comment (mikenorgate) [reply scripts/benchmark/ferment-acceptance.mjs]: Obsolete — the benchmark script this comment targets was removed from the branch in ea5a7efc (internal tooling, not shipped with the PR). The lifecycle/timeout/traversal concerns it raised no longer apply to the diff.

File: scripts/benchmark/ferment-acceptance.mjs [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+	return new Promise((resolvePromise) => {
+		server.listen(port, () => resolvePromise({ server, url: `http://localhost:${port}` }))
+	})
+}
+
+function startDevServer(artifactDir, pkg, port) {
+	const script = pkg.scripts?.dev ? "dev" : "preview"
+	const proc = spawn("npm", ["run", script, "--", "--port", String(port), "--strictPort"], {
+		cwd: artifactDir,
+		stdio: ["ignore", "pipe", "pipe"],
```
[the remark this answers is not in the document]
Comment (mikenorgate) [reply scripts/benchmark/ferment-acceptance.mjs]: Obsolete — the benchmark script this comment targets was removed from the branch in ea5a7efc (internal tooling, not shipped with the PR). The lifecycle/timeout/traversal concerns it raised no longer apply to the diff.

File: scripts/benchmark/ferment-acceptance.mjs [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+	if (args.playwright) candidates.push(args.playwright)
+	candidates.push(join(args.artifactDir, "package.json"), join(import.meta.dirname, "../../package.json"))
+	for (const base of candidates) {
+		try {
+			const req = createRequire(base)
+			return req("playwright")
+		} catch {}
+	}
+	throw new Error("could not resolve playwright — pass --playwright <absolute path to its package dir>")
+}
```
[the remark this answers is not in the document]
Comment (mikenorgate) [reply scripts/benchmark/ferment-acceptance.mjs]: Obsolete — the benchmark script this comment targets was removed from the branch in ea5a7efc (internal tooling, not shipped with the PR). The lifecycle/timeout/traversal concerns it raised no longer apply to the diff.

File: scripts/benchmark/ferment-acceptance.mjs [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+	})
+}
+
+function startDevServer(artifactDir, pkg, port) {
+	const script = pkg.scripts?.dev ? "dev" : "preview"
+	const proc = spawn("npm", ["run", script, "--", "--port", String(port), "--strictPort"], {
+		cwd: artifactDir,
+		stdio: ["ignore", "pipe", "pipe"],
+	})
+	proc.stdout.on("data", () => {})
```
[the remark this answers is not in the document]
Comment (mikenorgate) [reply scripts/benchmark/ferment-acceptance.mjs]: Obsolete — the benchmark script this comment targets was removed from the branch in ea5a7efc (internal tooling, not shipped with the PR). The lifecycle/timeout/traversal concerns it raised no longer apply to the diff.

File: scripts/benchmark/ferment-acceptance.mjs [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+			// server not up yet — expected during boot; keep polling
+		}
+		if (Date.now() > deadline) throw new Error(`server at ${url} did not become ready within ${timeoutMs}ms`)
+		await new Promise((r) => setTimeout(r, 250))
+	}
+}
+
+function resolvePlaywright(args) {
+	const candidates = []
+	if (args.playwright) candidates.push(args.playwright)
```
[the remark this answers is not in the document]
Comment (mikenorgate) [reply scripts/benchmark/ferment-acceptance.mjs]: Obsolete — the benchmark script this comment targets was removed from the branch in ea5a7efc (internal tooling, not shipped with the PR). The lifecycle/timeout/traversal concerns it raised no longer apply to the diff.

File: scripts/benchmark/ferment-acceptance.mjs [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+	await browser.close()
+	writeFileSync(join(args.out, "probe.json"), JSON.stringify(probe, null, 2))
+
+	if (proc) proc.kill("SIGTERM")
+	if (server) server.close()
+
+	const manifest = {
+		label: args.label,
+		artifactDir: args.artifactDir,
+		harvestedAt: new Date().toISOString(),
```
[the remark this answers is not in the document]
Comment (mikenorgate) [reply scripts/benchmark/ferment-acceptance.mjs]: Obsolete — the benchmark script this comment targets was removed from the branch in ea5a7efc (internal tooling, not shipped with the PR). The lifecycle/timeout/traversal concerns it raised no longer apply to the diff.

File: src/extensions/ferment/judge.ts [hunk 7]
```diff
@@ -406,7 +447,20 @@ export async function judgeJourneyGrade(
 		}
```
[the remark this answers is not in the document]
Comment (mikenorgate) [reply src/extensions/ferment/judge.ts:458]: Fixed in 6d2107a5 — the retry now rebuilds the user message with an explicit reminder naming the omitted charter_verdicts and the required shape (one entry per clause with met/waived/unmet + evidence). Regression test asserts the reminder lands on the retry attempt only.

File: src/extensions/ferment/auto-compaction.ts [hunk 8]
```diff
@@ -316,6 +330,15 @@ export function buildMidTurnCustomInstructions(
...
 	}
 
+	if (ferment.charter) {
+		lines.push("- Intent charter (preserve verbatim — it is the grading anchor for this ferment):")
+		for (const line of renderCharterCompact(ferment.charter).split("\n")) lines.push(`  ${line}`)
+	}
+
+	lines.push(
+		"- Open concerns: preserve any open concerns, known gaps, deferred fixes, or quality worries (things noticed but unresolved) — they must survive compaction alongside the plan. Do not drop one merely because it has no owner or fix yet.",
+	)
```
Comment (pmateusz) [inline src/extensions/ferment/auto-compaction.ts]: nit: we are duplicating this code fragment (same fragment also exists 269-277). Helper function to render a prompt section / append prompt lines will be helpful to prevent drift.

File: src/extensions/ferment/judge.ts [hunk 9]
```diff
@@ -40,6 +41,24 @@ export type JudgeUnavailableReason = "no_registry" | "no_model" | "no_auth" | "a
...
 export type JudgeApiResult = { ok: true; text: string } | { ok: false; reason: JudgeUnavailableReason; detail?: string }
 
+/**
+ * Resolve the judge model's display ref for observability (mirrors
+ * judgeApiCall's resolution order): the model-roles `judge` assignment when it
+ * resolves in the captured registry, otherwise the session-model fallback
+ * captured at extension init. Returns `provider/id`, or undefined when neither
+ * side is known (unit tests that inject apiCall hit this).
+ */
+export function describeJudgeModel(): string | undefined {
```
Comment (pmateusz) [inline src/extensions/ferment/judge.ts:63]: bug: `gradedBy` can record a model that did not produce the grade. The production path tries the Grader subagent first, but `spawnGraderAgent` calls `spawnAndWait(..., "Grader", ...)` without a `model`. The agent runner therefore falls back to `ctx.model` (the parent session model), while this helper independently resolves the configured `modelRoles.judge`. When those models differ, a successful subagent grade is persisted and rendered with incorrect provenance.

The example below is not a real regression test, but it shows that looking up model configuration doesn't cover all edge cases.

```ts
it("spawns the Grader with the configured judge model rather than the parent session model", async () => {
  const judgeModel = {
    id: "judge-model",
    provider: "kimchi-dev",
    name: "Judge",
    input: ["text"],
  }
  const parentModel = {
    id: "parent-model",
    provider: "kimchi-dev",
    name: "Parent",
    input: ["text"],
  }
  const ctx = makeMockCtx(
    makeMockModelRegistry([judgeModel, parentModel]),
    parentModel,
  )

  await spawnGraderAgent(pi, ctx, "grade this ferment")

  expect(manager.spawnAndWait).toHaveBeenCalledWith(
    pi,
    ctx,
    "Grader",
    "grade this ferment",
    expect.objectContaining({ model: judgeModel }),
  )
})
```

The failure shows the received spawn options contain `description`, `visibility`, session paths, and `signal`, but no `model`. Please either resolve and pass the judge model into the Grader spawn, or return the actual model provenance from the grading path and persist that value.

**Review by pmateusz — COMMENTED** [review]
Besides testing on your coding example, did you have a chance to obtain a benchmark run? I doubt the benchmark examples will exercise your change, though they could be helpful to capture possible regressions.

**Comment by mikenorgate** [comment]
> Besides testing on your coding example, did you have a chance to obtain a benchmark run? I doubt the benchmark examples will exercise your change, though they could be helpful to capture possible regressions.

Yes did a benchmark run, got slightly less fails compared to master but had a few more timeouts so score came out about the same

File: src/extensions/ferment/judge.ts (hunk 9 above)
Comment (mikenorgate) [reply src/extensions/ferment/judge.ts:63]: Fixed — the Grader subagent now runs on the configured judge-role model. resolveGraderModel() mirrors judgeApiCall/describeJudgeModel (roles.judge ref → registry.find → undefined fallback) and passes the resolved model into spawnAndWait, so gradedBy provenance matches the model that actually graded. When the role doesn't resolve, the runner falls back to ctx.model, which is the same model describeJudgeModel reports. Regression test in agents/index.test.ts controls the judge role via the test-env model-roles mock and asserts the model lands in the spawn options (plus the fallback-omits-model case). Commits on this branch.

File: src/extensions/ferment/auto-compaction.ts (hunk 8 above)
Comment (mikenorgate) [reply src/extensions/ferment/auto-compaction.ts]: Done — appendPlanAnchorLines() now owns the ferment name/goal, success criteria, intent charter, and open-concerns block once, and both buildCustomInstructions and buildMidTurnCustomInstructions call it. The charter preservation wording especially can no longer drift between the two compaction paths.

**Review by pmateusz — APPROVED** [review]
